cylc/flow/task_events_mgr.py

```diff
@@ -1284,6 +1284,13 @@ def _insert_task_job(
         # not see previous submissions (so can't use itask.jobs[submit_num-1]).
         job_conf = itask.jobs[-1]
 
+        # Extract a resolved platform name if one is available from the itask
+        # else get an unresolved platform name from the task def:
+        if itask.platform:
+            platform_name = itask.platform['name']
+        else:
+            platform_name = itask.tdef.rtconfig['platform']
```

Comment:
In what case was `itask.platform` not set?

Logically the platform must have been set before the job was submission was made.

Reply:
When it's called by `_process_message_submit_failed`.

Reply:
The platform gets set during submission:

https://github.com/cylc/cylc-flow/blob/a52943a034b195c08ffd0195636a80d94e7401b1/cylc/flow/task_job_mgr.py#L279-L280

So it should be set for submit-failed jobs. This functionality is tested so it definitely works.

Have you identified a scenario where the submission fails before the platform is set there?

Reply:
Proved to myself that the platform is set with this example:

```diff
diff --git a/cylc/flow/task_events_mgr.py b/cylc/flow/task_events_mgr.py
index ea2c7c1be..3b177cb50 100644
--- a/cylc/flow/task_events_mgr.py
+++ b/cylc/flow/task_events_mgr.py
@@ -1207,6 +1207,7 @@ class TaskEventsManager():
 
         # Register newly submit-failed job with the database and datastore.
         job_d = itask.tokens.duplicate(job=str(itask.submit_num)).relative_id
+        print('#', itask.platform)
         self._insert_task_job(itask, event_time, self.JOB_SUBMIT_FAIL_FLAG)
         self.data_store_mgr.delta_job_state(job_d, TASK_STATUS_SUBMIT_FAILED)
 ```

```
[scheduling]    
    [[graph]]
        R1 = foo

[runtime]
    [[foo]]
        platform = foo
```

```console
(cylc8) [cylc-run] $ cylc play submit-failed --no-detach

 ▪ ■  Cylc Workflow Engine 8.1.1.dev
 ██   Copyright (C) 2008-2023 NIWA
▝▘    & British Crown (Met Office) & Contributors

...

2023-01-25T15:52:37Z INFO - [1/foo waiting(runahead) job:00 flows:1] => waiting
2023-01-25T15:52:37Z INFO - [1/foo waiting job:00 flows:1] => waiting(queued)
2023-01-25T15:52:37Z INFO - [1/foo waiting(queued) job:00 flows:1] => waiting
2023-01-25T15:52:37Z INFO - [1/foo waiting job:01 flows:1] => preparing
2023-01-25T15:52:37Z ERROR - [jobs-submit cmd] (platform not defined)
    [jobs-submit ret_code] 1
    [jobs-submit err] No matching platform "foo" found
2023-01-25T15:52:37Z CRITICAL - [1/foo preparing job:01 flows:1] submission
    failed
2023-01-25T15:52:37Z INFO - [1/foo preparing job:01 flows:1] => submit-failed
# OrderedDictWithDefaults([('meta', OrderedDictWithDefaults()), ('hosts', ['localhost']), ('job runner', 'background'), ('job runner command template', None), ('shell', '/bin/bash'), ('communication method', 'zmq'), ('submission polling intervals', [900.0]), ('submission retry delays', None), ('execution polling intervals', [900.0]), ('execution time limit polling intervals', [60.0, 120.0, 420.0]), ('ssh command', 'ssh -oBatchMode=yes -oConnectTimeout=10'), ('rsync command', 'rsync'), ('use login shell', True), ('cylc path', None), ('global init-script', None), ('copyable environment variables', ''), ('retrieve job logs', None), ('retrieve job logs command', 'rsync -a'), ('retrieve job logs max size', None), ('retrieve job logs retry delays', []), ('tail command template', 'tail -n +1 -F %(filename)s'), ('err tailer', None), ('out tailer', None), ('err viewer', None), ('out viewer', None), ('job name length maximum', None), ('install target', None), ('clean job submission environment', False), ('job submission environment pass-through', []), ('job submission executable paths', []), ('max batch submit size', 100), ('selection', OrderedDictWithDefaults([('method', 'definition order')])), ('directives', OrderedDictWithDefaults()), ('name', 'localhost')])
2023-01-25T15:52:37Z ERROR - Incomplete tasks:
      * 1/foo did not complete required outputs: ['succeeded']
2023-01-25T15:52:37Z CRITICAL - Workflow stalled
2023-01-25T15:52:37Z WARNING - PT1H stall timer starts NOW
```

Reply:
With your example, and your global config I get this total failure:

```
2023-01-25T16:16:16Z ERROR - name 'platform_name' is not defined
    Traceback (most recent call last):
      File "/net/home/h02/tpilling/metomi/cylc-flow/cylc/flow/task_job_mgr.py", line 1165, in _prep_submit_task_job
        platform = get_platform(
      File "/net/home/h02/tpilling/metomi/cylc-flow/cylc/flow/platforms.py", line 143, in get_platform
        return platform_from_name(task_conf['platform'], bad_hosts=bad_hosts)
      File "/net/home/h02/tpilling/metomi/cylc-flow/cylc/flow/platforms.py", line 249, in platform_from_name
        raise PlatformLookupError(
    cylc.flow.exceptions.PlatformLookupError: No matching platform "foo" found
    During handling of the above exception, another exception occurred:
    Traceback (most recent call last):
      File "/net/home/h02/tpilling/metomi/cylc-flow/cylc/flow/scheduler.py", line 634, in run_scheduler
        await self.main_loop()
      File "/net/home/h02/tpilling/metomi/cylc-flow/cylc/flow/scheduler.py", line 1617, in main_loop
        self.release_queued_tasks()
      File "/net/home/h02/tpilling/metomi/cylc-flow/cylc/flow/scheduler.py", line 1334, in release_queued_tasks
        for itask in self.task_job_mgr.submit_task_jobs(
      File "/net/home/h02/tpilling/metomi/cylc-flow/cylc/flow/task_job_mgr.py", line 270, in submit_task_jobs
        prepared_tasks, bad_tasks = self.prep_submit_task_jobs(
      File "/net/home/h02/tpilling/metomi/cylc-flow/cylc/flow/task_job_mgr.py", line 242, in prep_submit_task_jobs
        prep_task = self._prep_submit_task_job(
      File "/net/home/h02/tpilling/metomi/cylc-flow/cylc/flow/task_job_mgr.py", line 1174, in _prep_submit_task_job
        self._prep_submit_task_job_error(
      File "/net/home/h02/tpilling/metomi/cylc-flow/cylc/flow/task_job_mgr.py", line 1247, in _prep_submit_task_job_error
        'platform_name': platform_name,
    NameError: name 'platform_name' is not defined

```

Reply:
On consideration of the following example:

```ini
#flow.cylc 
[runtime]
    [[mytask]]
        platform = foobarbaz
```

```ini
# global.cylc
[platforms]
        [[platform1]]
            ... some settings

# Definitely no regex matching "foo bar baz"
```

When mytask fails to submit because no matching platform can be found - how does the platform definition get populated?

Reply:
>> Have you identified a scenario where the submission fails before the platform is set there?
>
> Yes - if submission fails because a platform cannot be selected.

I think this is correct, if the job failed to submit then it doesn't have a platform.

Reply:
Put back - lets see what happens on testing!